subsys/net/l2/canbus/canbus.c

```diff
@@ -1,49 +1,1211 @@
+{
+	k_mutex_lock(&rx_ctx_mtx, K_FOREVER);
+	ctx->state = NET_CAN_RX_STATE_UNUSED;
+	k_mutex_unlock(&rx_ctx_mtx);
+}
+
+static void tx_finish(struct net_pkt *pkt)
+{
+	struct can_net_isotp_tx_ctx *ctx = pkt->can_tx_ctx;
+
+	if (ctx->state != NET_CAN_TX_STATE_RESET) {
+		z_abort_timeout(&ctx->timeout);
+	}
+
+	free_tx_ctx(ctx);
+	net_pkt_unref(pkt);
+	k_sem_give(&tx_sem);
+}
+
+static void rx_finish(struct net_pkt *pkt)
 {
-	net_pkt_lladdr_src(pkt)->addr = NULL;
-	net_pkt_lladdr_src(pkt)->len = 0U;
-	net_pkt_lladdr_src(pkt)->type = NET_LINK_CANBUS;
-	net_pkt_lladdr_dst(pkt)->addr = NULL;
-	net_pkt_lladdr_dst(pkt)->len = 0U;
-	net_pkt_lladdr_dst(pkt)->type = NET_LINK_CANBUS;
+	struct can_net_isotp_rx_ctx *ctx = pkt->can_rx_ctx;
 
-	net_pkt_set_family(pkt, AF_CAN);
+	free_rx_ctx(ctx);
+}
+
+static void tx_report_err(struct net_pkt *pkt)
+{
+	tx_finish(pkt);
+}
+
+static void rx_report_err(struct net_pkt *pkt)
+{
+	rx_finish(pkt);
+	net_pkt_unref(pkt);
+}
+
+static void rx_err_work_handler(struct k_work *item)
+{
+	struct net_pkt *pkt = CONTAINER_OF(item, struct net_pkt, work);
+
+	rx_report_err(pkt);
+}
+
+static void rx_report_err_from_isr(struct net_pkt *pkt)
+{
+	k_work_init(&pkt->work, rx_err_work_handler);
+	k_work_submit_to_queue(&net_can_workq, &pkt->work);
+}
+
+static void tx_timeout(struct _timeout *t)
+{
+	struct can_net_isotp_tx_ctx *ctx =
+		CONTAINER_OF(t, struct can_net_isotp_tx_ctx, timeout);
+	NET_ERR("TX Timeout. CTX: %p", ctx);
+	ctx->state = NET_CAN_TX_STATE_ERR;
+	k_work_submit_to_queue(&net_can_workq, &ctx->pkt->work);
+}
+
+
+static void rx_timeout(struct _timeout *t)
+{
+	struct can_net_isotp_rx_ctx *ctx =
+		CONTAINER_OF(t, struct can_net_isotp_rx_ctx, timeout);
+	NET_ERR("RX Timeout. CTX: %p", ctx);
+	ctx->state = NET_CAN_RX_STATE_TIMEOUT;
+	rx_report_err_from_isr(ctx->pkt);
+}
+
+static void st_min_timeout(struct _timeout *t)
+{
+	struct can_net_isotp_tx_ctx *ctx =
+		CONTAINER_OF(t, struct can_net_isotp_tx_ctx, timeout);
+	k_work_submit_to_queue(&net_can_workq, &ctx->pkt->work);
+}
+
+static s32_t stmin_to_ticks(u8_t stmin)
+{
+	s32_t time_ms;
+	/* According to ISO 15765-2 stmin should be 127ms if value is corrupt */
+	if (stmin > NET_CAN_STMIN_MAX ||
+	    (stmin > NET_CAN_STMIN_MS_MAX && stmin < NET_CAN_STMIN_US_BEGIN)) {
+		time_ms = K_MSEC(NET_CAN_STMIN_MS_MAX);
+	} else if (stmin >= NET_CAN_STMIN_US_BEGIN) {
+		/* This should be 100us-900us but zephyr can't handle that*/
+		time_ms = K_MSEC(1);
+	} else {
+		time_ms = stmin;
+	}
+
+	return z_ms_to_ticks(time_ms);
+}
+
+static inline void cpy_lladdr(struct net_pkt *dst, struct net_pkt *src)
+{
+	struct net_buf *buf = dst->buffer;
+
+	net_pkt_lladdr_src(dst)->addr =  buf->data;
+	net_pkt_lladdr_src(dst)->len = net_pkt_lladdr_src(src)->len;
+	net_pkt_lladdr_src(dst)->type = net_pkt_lladdr_src(src)->type;
+	net_buf_add_mem(buf, net_pkt_lladdr_src(src)->addr,
+			net_pkt_lladdr_src(src)->len);
+	net_buf_pull(buf, net_pkt_lladdr_src(src)->len);
```

Comment:
Same comment to here for the net_buf_pull(), to me it looks like a wrong thing to do.